enhancements/microshift/y-2-upgrades.md

```diff
@@ -0,0 +1,223 @@
+  to update storage versions of any resources.
+* There are not currently any changes to the etcd storage format.
+* The version skew check in MicroShift itself is straightforward to
+  change.
+
+### Workflow Description
+
+1. Edge device administrator deploys a host with MicroShift 4.Y-2
+   installed.
+2. Software runs, time passes.
+3. Edge device administrator updates the host to run MicroShift 4.Y.
+  * For ostree-based systems, the host is automatically rebooted as
+    part of the update process.
+  * For RPM-based systems, the user must reboot the host after the
+    software update is completed.
+4. Edge device restarts.
+5. MicroShift restarts.
+6. MicroShift checks the data and binary version difference for
+   compatibility.
+7. If the check fails, MicroShift exits with an error.
+8. If the check passes, MicroShift continues to run, including
+   performing any data migration necessary.
+
+### API Extensions
+
+N/A
+
+### Risks and Mitigations
+
+There is a risk that some underlying data format will change between
+MicroShift versions (kubernetes storage versions, etcd file format,
+etc.). If that happens, we will have to build a tool to support
+migrating from 4.Y-2 to 4.Y-1 _anyway_. We will need to carry over the
+use of that tool for an extra release to support the 2 version upgrade
+capability.
+
+If we extend the supported upgrade skew, we would have to continue to
+carry the migration tool for the full length of the allowed upgrade
+window after 4.Y-1 (if the allowed skew is 5, we would carry the tool
+in 4.Y-1, 4.Y, 4.Y+1, 4.Y+2, and 4.Y+3 to support upgrading 4.Y-1 to
+4.Y+3 at one time).
+
+The [kubernetes version skew
+policy](https://kubernetes.io/releases/version-skew-policy/) is
+written assuming multi-node clusters. Even so, it supports 3
+kubernetes version difference between the API server and kubelet and 1
+version between the API server instances. This is what allows
+OpenShift's EUS upgrade process, in which the control plane is updated
+independently of the worker nodes, to work. In a single-node
+MicroShift deployment, the API server and kubelet are in the same
+binary and have the same version, so there is no skew at all.
+
+If, in the future, MicroShift does need to support multi-node
+deployments there will be many other aspects of deployment and upgrade
+to consider, in addition to the version skew problem. We can envision
+implementing a process similar to what OpenShift uses, where the
+control plane and workers are updated using separate steps.
+
+If an upgrade fails, even after a complex data migration, MicroShift's
+rollback process is to discard the new database and restore the old
+version from a backup before continuing. This ensures that an old
+version of the software matches the older database (file format,
+schema, and content).
+
+### Drawbacks
+
+The main drawback to implementing this enhancement is the increased
+test matrix for upgrades. We can automate those tests to minimize the
+impact.
```

Comment:
Right! So QE needs to be involved. I am waiting for feedback from them.

Reply:
1. The scope of QE testing will be confined to upgrades of RHEL and rpm-ostree hosts deployed with RPMs built by ART (nightly, EC, RC and after release z-streams)
2. Both x86_64 and aarch64 architecture will be covered
    a. Not every combination listed below will be tested on both architectures
    b. A mix of Intel and ARM hosts will be used
3. Rpm-ostree based hosts
    a. Initial cluster bring-up will be a mix of deployments from ISO installer and rpm-ostree upgrades from a bare RHEL host
    b. The following upgrade paths will be covered:
        i. RHEL 9.2 / 4.14.latest → RHEL 9.4 / 4.16.Z
       ii. RHEL 9.3 / 4.14.latest → RHEL 9.4 / 4.16.Z
      iii. RHEL 9.2 / 4.15.latest → RHEL 9.4 / 4.16.Z
      iv. RHEL 9.3 / 4.15.latest → RHEL 9.4 / 4.16.Z
    c. Z-stream upgrade testing for 4.16 will also include Z-1
        i. RHEL 9.4 / 4.16.Z-1 → RHEL 9.4 / 4.16.Z
    d. Rollback of each upgrade listed above will be performed
    e. Z-stream upgrade testing for 4.14 and 4.15 will remain at Y-1 and Z-1
4. RPM upgrades on RHEL hosts
    a. Given the support statement[1] says MicroShift will support only “2 RHEL versions: the most recent EUS at the time of development and the next RHEL release after that EUS release”, upgrade testing of MicroShift versions 4.14 or 4.15 to 4.16 will also require an OS upgrade from 9.2 or 9.3 to 9.4.
        i. MicroShift will need to be stopped and upgraded at the same time
            1. MicroShift 4.14 or 4.15 are not supported on RHEL 9.4
            2. MicroShift 4.16 is not supported on RHEL 9.2 or 9.3
    b. The upgrade paths described above will also be covered for RPM installed on RHEL
    c. RHEL states that it is unsafe to downgrade to an earlier minor release or an earlier EUS repository. Therefore, downgrades of MicroShift deployed with RPMs on RHEL is not supported and will not be tested

Reply:
> 1. The scope of QE testing will be confined to upgrades of RHEL and rpm-ostree hosts deployed with RPMs built by ART (nightly, EC, RC and after release z-streams)
> 2. Both x86_64 and aarch64 architecture will be covered
>    a. Not every combination listed below will be tested on both architectures
>    b. A mix of Intel and ARM hosts will be used

That seems fine, we don't really expect the architecture to affect upgrades in a way that is unique to the versions.

> 3. Rpm-ostree based hosts
>    a. Initial cluster bring-up will be a mix of deployments from ISO installer and rpm-ostree upgrades from a bare RHEL host
>    b. The following upgrade paths will be covered:
>    i. RHEL 9.2 / 4.14.latest → RHEL 9.4 / 4.16.Z
>    ii. RHEL 9.3 / 4.14.latest → RHEL 9.4 / 4.16.Z
>    iii. RHEL 9.2 / 4.15.latest → RHEL 9.4 / 4.16.Z
>    iv. RHEL 9.3 / 4.15.latest → RHEL 9.4 / 4.16.Z

Since we don't expect the OS to play a part in EUS updates working or failing, if you wanted to shrink that list you could safely limit the combinations to 9.2/4.14->9.4/4.16 and 9.3/4.15->9.4/4.16. If you have capacity to run all 4 tests, that's good, too.

+1 for the plan generally, thanks @jogeo !

Reply:
I added that test plan to the doc, @jogeo